crates/autopilot/src/infra/persistence/mod.rs

```diff
@@ -542,12 +494,46 @@ impl Persistence {
             !expired && !invalidated && !onchain_error && !fulfilled
         });
 
-        // Keep only relevant quotes.
-        next_quotes.retain(|uid, _quote| current_orders.contains_key(uid));
+        let new_quotes: HashMap<_, _> = {
+            let _timer = Metrics::get()
+                .database_queries
+                .with_label_values(&["read_quotes"])
+                .start_timer();
+
+            // Fetch quotes for only newly created and placed onchain orders.
+            let order_uids = current_orders
+                .values()
+                .filter_map(|order| {
+                    (order.metadata.onchain_user.is_some()
+                        || order.metadata.creation_date > after_timestamp)
+                        .then_some(ByteArray(order.metadata.uid.0))
+                })
+                .collect::<Vec<_>>();
+
+            database::orders::read_quotes(&mut tx, &order_uids)
+                .await?
+                .into_iter()
+                .filter_map(|quote| {
+                    let order_uid = domain::OrderUid(quote.order_uid.0);
+                    dto::quote::into_domain(quote)
+                        .map_err(|err| {
+                            tracing::warn!(?order_uid, ?err, "failed to convert quote from db")
+                        })
+                        .ok()
+                        .map(|quote| (order_uid, quote))
+                })
+                .collect()
+        };
+
+        current_quotes.retain(|uid, _| current_orders.contains_key(uid));
+
+        for (order_uid, quote) in new_quotes {
+            current_quotes.insert(order_uid, quote);
+        }
```

Comment:
We could avoid this loop and `new_quotes` temp variable by:
1. moving `current_quotes.retain()` before quote fetching block
2. using `current_quotes.extend(new_quotes_iter)` (to avoid collecting the transformed quotes)